frontend/src/pages/projects/screens/spawner/useNotebookSizeState.ts

```diff
@@ -0,0 +1,45 @@
+import React from 'react';
+import useNotebookDeploymentSize from '~/pages/projects/screens/detail/notebooks/useNotebookDeploymentSize';
+import { NotebookKind } from '~/k8sTypes';
+import { NotebookSize } from '~/types';
+import { getCustomNotebookSize } from '~/pages/projects/utils';
+import { useNotebookSize } from './useNotebookSize';
+
+export const useNotebookSizeState = (
+  existingNotebook: NotebookKind | undefined,
+): {
+  selectedSize: NotebookSize;
+  setSelectedSize: (size: string) => void;
+  sizes: NotebookSize[];
+} => {
+  const { size: notebookSize } = useNotebookDeploymentSize(existingNotebook);
+  const { sizes } = useNotebookSize();
+
+  const allSizes = React.useMemo(() => {
+    if (notebookSize) {
+      return sizes;
+    }
+    const CUSTOM_NOTEBOOK_SIZE = getCustomNotebookSize(existingNotebook);
+    if (existingNotebook && !sizes.some((size) => size.name === CUSTOM_NOTEBOOK_SIZE.name)) {
+      return [CUSTOM_NOTEBOOK_SIZE, ...sizes];
+    }
+    return sizes;
+  }, [notebookSize, sizes, existingNotebook]);
+
+  const defaultSize = React.useMemo(() => {
+    const foundSize = allSizes.find((size) => size.name === notebookSize?.name);
+    return foundSize || allSizes[0];
+  }, [allSizes, notebookSize]);
+
+  const [selectedSize, setSelectedSize] = React.useState<NotebookSize>(defaultSize);
```

Comment:
Use the `useState` function initializer param since the value of `defaultSize` is only used once to initialize the state.

```suggestion
  const [selectedSize, setSelectedSize] = React.useState<NotebookSize>(
    () => allSizes.find((size) => size.name === notebookSize?.name) || allSizes[0],
  );
```

Reply:
TIL you can do that 👀

Reply:
done, thanks for pointing this out.